### PR TITLE
Separate Layout from Data

### PR DESCRIPTION
**Pull request by connormanning**
This branch separates the layout (dimension information - types/sizes/offsets etc.) from the physical point data.  The purpose is entirely for API clarity and usage, since this separation isn't important to PDAL's internal workings.

### Previous
 - `PointContext` - Holds the buffer of raw point data (via member RawPtBuf) as well as layout information (via DimInfo).  Layout/data information are not conceptually separable from this member.
 - `PointBuffer` - A 'view' into a PointContext made of indices into the Context's point data.  Allows for simple access into the PointContext's raw point data (via set/getField), and allows for fast filtering/indexing operations by rearranging indices into the PointContext's data buffer, avoiding data copying.

### Proposed
 - `PointLayout` - Direct copy of the *layout* functionality previously contained in PointContext.  No points contained here - this is just what points using this layout *look* like.
 - `PointTable` - Direct copy of the *data* functionality previously contained in PointContext via what used to be called `RawPtBuf`.  A `PointTable` contains a reference member to a PointLayout, which describes how its data is laid out.  This is where the actual point data lives.  This is virtual, so you can change the physical storage as a client.  PDAL's version uses a set of vectors, which are preallocated.
 - `PointView` - A rename of `PointBuffer` with no functional change, described above.

All these structures use shared_ptr types to maintain consistency with the rest of PDAL.

### Effects
What would this mean to the PDAL codebase?  Almost (but not quite) nothing, as these changes are mostly conceptual:
 - Where you used to say `PointContext ctx`, now you have to say `PointTablePtr table(new DefaultPointTable())`, and use the resulting `table` wherever you previously used `ctx`.
 - To access layout information to a `PointTable`, there is an explicit indirection through its layout member.  A previous `PointBuffer` user doesn't see this change due to existing pass-throughs to this information from a `PointBuffer/View`.
 - `s/PointContext/PointTablePtr/` and `s/PointBuffer/PointView/`.
 - `Stage::addDimensions` now gives you a Table's `PointLayout` instead of the `PointTable` directly since this is a layout-only operation.

### Notes
There is *no new functionality* to internal PDAL itself.  The API becomes more powerful, and in a way that I personally think is more intuitive (arguments?).  It allows a libpdal user to more clearly separate the concerns of layout/data, where previously these two were married together in a `PointContext`.  This separation has proven important to me as an API user.

The header files for PointLayout, PointTable, and PointView are the most important part of the changes.  The implementation files are largely identical to their previous counterparts with minor tweaks to support the architecture change.  The rest of the diff is almost entirely renaming noise.  Looking at a few of the Reader/Filter diffs should be indicative of the minimal impact on the API at large.

A PointTable still has access to everything a PointContext used to have.  A PointView is identical to a PointBuffer, which didn't actually contain the buffer of points (it was named this way for legacy reasons).

Comments/questions/concerns welcome.

### CODE REVIEW

**Comment by abellgithub** [comment]
I think I'm OK with this.  My objection is really that there is a ton of change here.  I know most of it is naming, but it's intermixed with the structural changes, and that makes it kind of difficult to review well.

Seems like a PointTable constructor should have default ctor that behaves as the old PointContext behaved.

File: include/pdal/PipelineManager.hpp [hunk 1]
```diff
@@ -49,7 +49,11 @@ class PDAL_DLL PipelineManager
 {
 public:
     PipelineManager()
```
Comment (abellgithub) [inline include/pdal/PipelineManager.hpp]: Why do we insist on this?  C++ has default constructors for a reason.

**Comment by abellgithub** [comment]
Any reason why we shouldn't run with this?

**Comment by connormanning** [comment]
Currently migrating the optional plugins since I hadn't been building everything when I initially ran through the codebase with the updates.  Any input on the changes from @chambbj, @mpgerlek, or any others?

**Comment by chambbj** [comment]
@connormanning: I suppose I can adapt with time. If you fix the PCL code for me that'll help ;)

**Comment by connormanning** [comment]
@chambjj the 69b1543 commit migrates the rest of the codebase to match this pull request, so hopefully impact on you is minimal.  Take a look at some of the PCL diffs in that commit to get a feel for the changes.

**Comment by chambbj** [comment]
@connormanning It _looks_ fine. I'll test shortly - don't anticipate any problems.

**Comment by chambbj** [comment]
Not all PCL-specific, but here's what I get on Windows (MSVC2012). https://gist.github.com/chambbj/d4d4377be163e5fcba16#file-gistfile1-txt

**Comment by hobu** [comment]
PointLayout not exported? PDAL_DLL might not be defined if pdal_export.hpp isn't actually included.

**Comment by chambbj** [comment]
Correct, it wasn't. Neither are `PointTable` and `PointViewIter`. Rebuilding to see if that takes care of everything.

**Comment by connormanning** [comment]
PointViewIter shouldn't need to be since that's still header-only.  PointTable/DefaultPointTable/PointLayout classes definitely need it - just added them.

**Comment by chambbj** [comment]
Yes, I missed `DefaultPointTable`. I think we're close.

**Comment by chambbj** [comment]
`s/addBuffer/addView` - there is at least one occurrence in SmoothKernel.cpp

File: plugins/pcl/io/PcdWriter.cpp [hunk 2]
```diff
@@ -79,11 +79,11 @@ Options PcdWriter::getDefaultOptions()
 }
 
 
-void PcdWriter::write(const PointBuffer& data)
+void PcdWriter::write(const PointViewPtr view)
 {
     pcl::PointCloud<XYZIRGBA>::Ptr cloud(new pcl::PointCloud<XYZIRGBA>);
     BOX3D const& buffer_bounds = data.calculateBounds();
```
Comment (chambbj) [inline plugins/pcl/io/PcdWriter.cpp]: `data` is now `view`

File: plugins/pcl/io/PCLVisualizer.cpp [hunk 3]
```diff
@@ -140,16 +140,16 @@ CREATE_SHARED_PLUGIN(1, 0, PclVisualizer, Writer, s_info)
...
-    BOX3D const& buffer_bounds = data.calculateBounds();
+    BOX3D const& buffer_bounds = view->calculateBounds();
 
     typedef XYZIRGBA PointType;
 
-    // Convert PointBuffer to a PCL PointCloud
+    // Convert PointView to a PCL PointCloud
     pcl::PointCloud<PointType>::Ptr cloud(new pcl::PointCloud<PointType>);
-    pclsupport::PDALtoPCD(const_cast<PointBuffer&>(data), *cloud, buffer_bounds);
+    pclsupport::PDALtoPCD(const_cast<PointViewPtr>(view), *cloud, buffer_bounds);
```
Comment (chambbj) [inline plugins/pcl/io/PCLVisualizer.cpp]: `Error	460	error C2440: 'const_cast' : cannot convert from 'const pdal::PointViewPtr' to 'pdal::PointViewPtr'`

File: plugins/pcl/filters/PCLBlock.cpp [hunk 4]
```diff
@@ -129,7 +129,7 @@ PointBufferSet PCLBlock::run(PointBufferPtr input)
     if (cloud_f->points.empty())
     {
         log()->get(LogLevel::Debug2) << "Filtered cloud has no points!" << std::endl;
-        return pbSet;
+        return viewSet;
     }
 
     pclsupport::PCDtoPDAL(*cloud_f, *output, buffer_bounds);
```
Comment (chambbj) [inline plugins/pcl/filters/PCLBlock.cpp]: `Error	474	error C2664: 'void pdal::pclsupport::PCDtoPDAL<T>(CloudT &,pdal::PointViewPtr,const pdal::BOX3D &)' : cannot convert parameter 2 from 'pdal::PointView' to 'pdal::PointViewPtr'`

File: plugins/pcl/filters/PCLBlock.cpp [hunk 5]
```diff
@@ -78,7 +78,7 @@ PointBufferSet PCLBlock::run(PointBufferPtr input)
 
     BOX3D const& buffer_bounds = input->calculateBounds();
 
-    // convert PointBuffer to PointNormal
+    // convert PointView to PointNormal
     typedef pcl::PointCloud<pcl::PointXYZ> Cloud;
     Cloud::Ptr cloud(new Cloud);
     pclsupport::PDALtoPCD(*input, *cloud, buffer_bounds);
```
Comment (chambbj) [inline plugins/pcl/filters/PCLBlock.cpp]: `Error	473	error C2664: 'void pdal::pclsupport::PDALtoPCD<T>(pdal::PointViewPtr,CloudT &,const pdal::BOX3D &)' : cannot convert parameter 1 from 'pdal::PointView' to 'pdal::PointViewPtr'`

File: plugins/pcl/io/PcdWriter.cpp [hunk 6]
```diff
@@ -79,11 +79,11 @@ Options PcdWriter::getDefaultOptions()
...
-    pclsupport::PDALtoPCD(const_cast<PointBuffer&>(data), *cloud, buffer_bounds);
+    pclsupport::PDALtoPCD(const_cast<PointViewPtr>(view), *cloud, buffer_bounds);
```
Comment (chambbj) [inline plugins/pcl/io/PcdWriter.cpp]: `Error	451	error C2440: 'const_cast' : cannot convert from 'const pdal::PointViewPtr' to 'pdal::PointViewPtr'`

File: plugins/nitf/test/NitfReaderTest.cpp [hunk 7]
```diff
@@ -166,10 +166,10 @@ TEST(NitfReaderTest, optionSrs)
...
 
-    writer.prepare(ctx);
-    PointBufferSet pbSet = writer.execute(ctx);
+    writer.prepare(table);
+    PointViewSet pbSet = writer.execute(table);
 
     EXPECT_EQ(sr, nitfReader->getSpatialReference().getWKT());
     EXPECT_EQ("", writer.getSpatialReference().getWKT());
-    EXPECT_EQ(sr, ctx.spatialRef().getWKT());
+    EXPECT_EQ(sr, table.spatialRef().getWKT());
```
Comment (chambbj) [inline plugins/nitf/test/NitfReaderTest.cpp]: `Error	842	error C2039: 'spatialRef' : is not a member of 'std::shared_ptr<_Ty>'`

**Comment by connormanning** [comment]
Closing in favor of a future PR to master.

**Pull request closed without merging**